File: leet-code/347. Top K Frequent Elements/TopKFrequentElements347_2.java
import java.util.*;

public class TopKFrequentElements347_2 {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }

        PriorityQueue<int[]> pq = new PriorityQueue<>((o1, o2) -> o2[1] - o1[1]);
        for (Integer i : map.keySet()) {
            pq.add(new int[]{i, map.get(i)});
        }
        int[] answer = new int[k];
        for (int i = 0; i < k; i++) {
            answer[i] = Objects.requireNonNull(pq.poll())[0];
        }
        return answer;
    }
}

File: leet-code/347. Top K Frequent Elements/TopKFrequentElements347_2.kt
package com.example

import java.util.PriorityQueue

class TopKFrequentElements347_2 {
    fun topKFrequent(nums: IntArray, k: Int): IntArray {
        val map = mutableMapOf<Int, Int>()
        nums.forEach { map[it] = map.getOrDefault(it, 0) + 1 }
        val pq = PriorityQueue<Element>(compareByDescending { it.value })
        map.forEach { entry -> pq.add(Element(entry.key, entry.value)) }
        val answer = IntArray(k)
        IntRange(1, k).forEachIndexed { i, _ ->  answer[i] = pq.poll()!!.key }
        return answer
    }

    private data class Element(val key: Int, val value: Int)
}
